{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "9f710dd5-5cf6-4272-a7f5-509f50a59840",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv \n",
    "import intervaltree\n",
    "from intervaltree import Interval, IntervalTree \n",
    "import pandas as pd\n",
    "import copy\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "261a3b0b-c59d-4a1f-be03-e72e361a3d05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_smallest_region(genome):\n",
    "    smallest=1000000000000000000\n",
    "    for i in genome:\n",
    "        for j in [-1,0,1]:\n",
    "            x=np.abs(list(genome[i])[j].end-list(genome[i])[j+1].begin)\n",
    "            if x<smallest:\n",
    "                smallest=x\n",
    "                print(i)\n",
    "    return smallest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ea82818b-de47-462b-916d-1400a809ce27",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name=\"/project/ReadStatistics/ML/telocentro_hg38.bed\"\n",
    "file=open(file_name,\"r\")\n",
    "data=file.readlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "5939c328-f069-470f-8874-c67d45a26b31",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome={}\n",
    "for i in data:\n",
    "    x=i.split(\"\\t\")\n",
    "    if x[0] not in genome.keys():\n",
    "        genome[x[0]]=IntervalTree()\n",
    "        genome[x[0]][int(x[1]):int(x[2])]=(int(x[1]),int(x[2]))\n",
    "    else:\n",
    "        genome[x[0]][int(x[1]):int(x[2])]=(int(x[1]),int(x[2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "7fa2adab-03b2-4f4f-9c86-bc05a603fb05",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in genome.keys():\n",
    "    genome[key].merge_overlaps(strict=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "0bd608df-2012-4ada-8797-e4a930f4dfa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "del genome[\"chrX\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "1173a4ba-beed-4d58-b120-d34dcfc4ec42",
   "metadata": {},
   "outputs": [],
   "source": [
    "del genome[\"chrY\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "c11cbf37-0233-4a08-9a79-81ac96a0a5f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Interval(0, 10000, (0, 10000))"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(genome[\"chr1\"])[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "55c06991-4dfd-41b0-afcb-7c56b601f32a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\n",
      "chr2\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr8\n",
      "chr10\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr18\n",
      "chr21\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "10890000"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_smallest_region(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "647635a4-86ba-4452-b33d-1c92dca3325b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr1': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(121700000, 125100000), Interval(248946422, 248956422, (248946422, 248956422))]),\n",
       " 'chr2': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(91800000, 96000000), Interval(242183529, 242193529, (242183529, 242193529))]),\n",
       " 'chr3': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(87800000, 94000000), Interval(198285559, 198295559, (198285559, 198295559))]),\n",
       " 'chr4': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(48200000, 51800000), Interval(190204555, 190214555, (190204555, 190214555))]),\n",
       " 'chr5': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(46100000, 51400000), Interval(181528259, 181538259, (181528259, 181538259))]),\n",
       " 'chr6': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(58500000, 62600000), Interval(170795979, 170805979, (170795979, 170805979))]),\n",
       " 'chr7': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(58100000, 62100000), Interval(159335973, 159345973, (159335973, 159345973))]),\n",
       " 'chr8': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(43200000, 47200000), Interval(145128636, 145138636, (145128636, 145138636))]),\n",
       " 'chr9': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(42200000, 45500000), Interval(138384717, 138394717, (138384717, 138394717))]),\n",
       " 'chr10': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(38000000, 41600000), Interval(133787422, 133797422, (133787422, 133797422))]),\n",
       " 'chr11': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(51000000, 55800000), Interval(135076622, 135086622, (135076622, 135086622))]),\n",
       " 'chr12': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(33200000, 37800000), Interval(133265309, 133275309, (133265309, 133275309))]),\n",
       " 'chr13': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(16500000, 18900000), Interval(114354328, 114364328, (114354328, 114364328))]),\n",
       " 'chr14': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(16100000, 18200000), Interval(107033718, 107043718, (107033718, 107043718))]),\n",
       " 'chr15': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(17500000, 20500000), Interval(101981189, 101991189, (101981189, 101991189))]),\n",
       " 'chr16': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(35300000, 38400000), Interval(90328345, 90338345, (90328345, 90338345))]),\n",
       " 'chr17': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(22700000, 27400000), Interval(83247441, 83257441, (83247441, 83257441))]),\n",
       " 'chr18': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(15400000, 21500000), Interval(80363285, 80373285, (80363285, 80373285))]),\n",
       " 'chr19': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(24200000, 28100000), Interval(58607616, 58617616, (58607616, 58617616))]),\n",
       " 'chr20': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(25700000, 30400000), Interval(64434167, 64444167, (64434167, 64444167))]),\n",
       " 'chr21': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(10900000, 13000000), Interval(46699983, 46709983, (46699983, 46709983))]),\n",
       " 'chr22': IntervalTree([Interval(0, 10000, (0, 10000)), Interval(13700000, 17400000), Interval(50808468, 50818468, (50808468, 50818468))])}"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "59a83ae0-586a-40c9-b3ff-85aafcaf0781",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MAKING SPLITS OF THE GENOME\n",
    "def get_splits(chromosome_name):\n",
    "    splits=IntervalTree()\n",
    "\n",
    "    x=sorted(genome[chromosome_name])\n",
    "    print(len(x))\n",
    "    split_length=(10**7)\n",
    "    for i in range(len(x)-1):\n",
    "        s=x[i].end\n",
    "        e=x[i+1].begin\n",
    "        for i in range(s+1,(s+e)//2,split_length+1):\n",
    "            splits[i:i+split_length]=(i,i+split_length)\n",
    "        for i in range( e-1 , (s+e)//2 + (e-s)%(split_length+1) , -split_length-1): #  -split_length-1 \n",
    "            splits[i-split_length:i]=(i-split_length,i)\n",
    "    return splits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "dea393ac-2e48-4455-9e93-2a2819908279",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n"
     ]
    }
   ],
   "source": [
    "splits=get_splits(\"chr21\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "647102c5-17c5-4d2a-b774-8ca245f55d86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(splits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "67be6b7e-cfee-48a6-9b43-2fc9a04bf981",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(splits)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "1b1c535b-184a-48fe-a66a-0c448a5edfd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "duplicated_regions=pd.read_csv(\"/project/ReadStatistics/ML/Full_Data/out_df_ws_jumps.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "4cab7e25-b29b-4475-8b9f-266cf61d5a24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>coor_s</th>\n",
       "      <th>coor_e</th>\n",
       "      <th>ids</th>\n",
       "      <th>jumps</th>\n",
       "      <th>length</th>\n",
       "      <th>centro</th>\n",
       "      <th>telo</th>\n",
       "      <th>gaps</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>MIR_s_r</th>\n",
       "      <th>Alu_s_r</th>\n",
       "      <th>Satellite_s_r</th>\n",
       "      <th>used_coor_l_s</th>\n",
       "      <th>used_coor_l_e</th>\n",
       "      <th>used_coor_r_s</th>\n",
       "      <th>used_coor_r_e</th>\n",
       "      <th>CG_frac_l</th>\n",
       "      <th>CG_frac_r</th>\n",
       "      <th>CG_frac_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10000</td>\n",
       "      <td>207666</td>\n",
       "      <td>id1</td>\n",
       "      <td>3</td>\n",
       "      <td>197666</td>\n",
       "      <td>121818793</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9950</td>\n",
       "      <td>10000</td>\n",
       "      <td>207666</td>\n",
       "      <td>207716</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>257666</td>\n",
       "      <td>297956</td>\n",
       "      <td>id2</td>\n",
       "      <td>3</td>\n",
       "      <td>40290</td>\n",
       "      <td>121728503</td>\n",
       "      <td>247666</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>257616</td>\n",
       "      <td>257666</td>\n",
       "      <td>297956</td>\n",
       "      <td>298006</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.538</td>\n",
       "      <td>0.391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>347968</td>\n",
       "      <td>535988</td>\n",
       "      <td>id3</td>\n",
       "      <td>1</td>\n",
       "      <td>188020</td>\n",
       "      <td>121490471</td>\n",
       "      <td>337968</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>347918</td>\n",
       "      <td>347968</td>\n",
       "      <td>535988</td>\n",
       "      <td>536038</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>585988</td>\n",
       "      <td>817292</td>\n",
       "      <td>id4</td>\n",
       "      <td>4</td>\n",
       "      <td>231304</td>\n",
       "      <td>121209167</td>\n",
       "      <td>575988</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>585938</td>\n",
       "      <td>585988</td>\n",
       "      <td>817292</td>\n",
       "      <td>817342</td>\n",
       "      <td>-1.000</td>\n",
       "      <td>0.412</td>\n",
       "      <td>0.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>817367</td>\n",
       "      <td>821400</td>\n",
       "      <td>id5</td>\n",
       "      <td>1</td>\n",
       "      <td>4033</td>\n",
       "      <td>121205059</td>\n",
       "      <td>807367</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>817317</td>\n",
       "      <td>817367</td>\n",
       "      <td>821400</td>\n",
       "      <td>821450</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6651</th>\n",
       "      <td>22</td>\n",
       "      <td>48911634</td>\n",
       "      <td>48912886</td>\n",
       "      <td>id4359</td>\n",
       "      <td>1</td>\n",
       "      <td>1252</td>\n",
       "      <td>33857316</td>\n",
       "      <td>1895582</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>48911584</td>\n",
       "      <td>48911634</td>\n",
       "      <td>48912886</td>\n",
       "      <td>48912936</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.451</td>\n",
       "      <td>0.468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6652</th>\n",
       "      <td>22</td>\n",
       "      <td>49383944</td>\n",
       "      <td>49385910</td>\n",
       "      <td>id4360</td>\n",
       "      <td>1</td>\n",
       "      <td>1966</td>\n",
       "      <td>34329626</td>\n",
       "      <td>1422558</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49383894</td>\n",
       "      <td>49383944</td>\n",
       "      <td>49385910</td>\n",
       "      <td>49385960</td>\n",
       "      <td>0.510</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6653</th>\n",
       "      <td>22</td>\n",
       "      <td>49386637</td>\n",
       "      <td>49388496</td>\n",
       "      <td>id4361</td>\n",
       "      <td>1</td>\n",
       "      <td>1859</td>\n",
       "      <td>34332319</td>\n",
       "      <td>1419972</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49386587</td>\n",
       "      <td>49386637</td>\n",
       "      <td>49388496</td>\n",
       "      <td>49388546</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6654</th>\n",
       "      <td>22</td>\n",
       "      <td>50432257</td>\n",
       "      <td>50442552</td>\n",
       "      <td>id4362</td>\n",
       "      <td>2</td>\n",
       "      <td>10295</td>\n",
       "      <td>35377939</td>\n",
       "      <td>365916</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>50432207</td>\n",
       "      <td>50432257</td>\n",
       "      <td>50442552</td>\n",
       "      <td>50442602</td>\n",
       "      <td>0.725</td>\n",
       "      <td>0.451</td>\n",
       "      <td>0.648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6655</th>\n",
       "      <td>22</td>\n",
       "      <td>50740515</td>\n",
       "      <td>50808468</td>\n",
       "      <td>id4363</td>\n",
       "      <td>2</td>\n",
       "      <td>67953</td>\n",
       "      <td>35686197</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>50740465</td>\n",
       "      <td>50740515</td>\n",
       "      <td>50808468</td>\n",
       "      <td>50808518</td>\n",
       "      <td>0.569</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6656 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    coor_s    coor_e     ids  jumps  length     centro     telo  \\\n",
       "0       1     10000    207666     id1      3  197666  121818793        0   \n",
       "1       1    257666    297956     id2      3   40290  121728503   247666   \n",
       "2       1    347968    535988     id3      1  188020  121490471   337968   \n",
       "3       1    585988    817292     id4      4  231304  121209167   575988   \n",
       "4       1    817367    821400     id5      1    4033  121205059   807367   \n",
       "...   ...       ...       ...     ...    ...     ...        ...      ...   \n",
       "6651   22  48911634  48912886  id4359      1    1252   33857316  1895582   \n",
       "6652   22  49383944  49385910  id4360      1    1966   34329626  1422558   \n",
       "6653   22  49386637  49388496  id4361      1    1859   34332319  1419972   \n",
       "6654   22  50432257  50442552  id4362      2   10295   35377939   365916   \n",
       "6655   22  50740515  50808468  id4363      2   67953   35686197        0   \n",
       "\n",
       "      gaps  genes  ...  MIR_s_r  Alu_s_r  Satellite_s_r  used_coor_l_s  \\\n",
       "0        2      6  ...        0        0              0           9950   \n",
       "1        2      0  ...        0        1              0         257616   \n",
       "2        2      1  ...        0        0              0         347918   \n",
       "3        1      2  ...        0        0              0         585938   \n",
       "4        0      1  ...        0        0              0         817317   \n",
       "...    ...    ...  ...      ...      ...            ...            ...   \n",
       "6651     0      0  ...        0        0              0       48911584   \n",
       "6652     0      0  ...        0        0              0       49383894   \n",
       "6653     0      0  ...        0        0              0       49386587   \n",
       "6654     0      1  ...        0        1              0       50432207   \n",
       "6655     1      3  ...        0        0              0       50740465   \n",
       "\n",
       "      used_coor_l_e  used_coor_r_s  used_coor_r_e  CG_frac_l  CG_frac_r  \\\n",
       "0             10000         207666         207716     -1.000      0.000   \n",
       "1            257666         297956         298006     -1.000      0.538   \n",
       "2            347968         535988         536038     -1.000      1.000   \n",
       "3            585988         817292         817342     -1.000      0.412   \n",
       "4            817367         821400         821450      0.392      0.529   \n",
       "...             ...            ...            ...        ...        ...   \n",
       "6651       48911634       48912886       48912936      0.490      0.451   \n",
       "6652       49383944       49385910       49385960      0.510      0.392   \n",
       "6653       49386637       49388496       49388546      0.392      0.314   \n",
       "6654       50432257       50442552       50442602      0.725      0.451   \n",
       "6655       50740515       50808468       50808518      0.569      0.000   \n",
       "\n",
       "      CG_frac_in  \n",
       "0          0.444  \n",
       "1          0.391  \n",
       "2          0.430  \n",
       "3          0.428  \n",
       "4          0.510  \n",
       "...          ...  \n",
       "6651       0.468  \n",
       "6652       0.393  \n",
       "6653       0.397  \n",
       "6654       0.648  \n",
       "6655       0.454  \n",
       "\n",
       "[6656 rows x 68 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicated_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "34dcc21c-23a8-4f5c-a43c-e1549f04b3fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "file=open(\"/project/ReadStatistics/ML/chr1.csv\",\"w\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "ae7eb1f8-bd50-48a8-9c00-9ec0cc37c014",
   "metadata": {},
   "outputs": [],
   "source": [
    "file.write(\"START\"+\",\"+\"END\"+\"\\n\")\n",
    "for i in sorted(splits):\n",
    "    file.write(str(i[0])+\",\"+str(i[1])+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "714a3f16-dada-444a-be45-702c1899e30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_duplicated_regions():\n",
    "    add_columns=['component_size', 'intra_degree', 'iner_degree', 'self_loops', 'edges_double', 'edges_tandem','edges_ident_mean']\n",
    "    old_data=pd.read_csv(\"/project/ReadStatistics/ML/df_Dinesh_data.csv\")\n",
    "    duplicated_regions=pd.read_csv(\"/project/ReadStatistics/ML/Full_Data/out_df_ws_jumps.csv\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    duplicated_regions[\"CG_frac_l\"][duplicated_regions[\"CG_frac_l\"]==-1.000]=0.0\n",
    "    duplicated_regions[\"CG_frac_l\"][duplicated_regions[\"CG_frac_l\"]==0.0]=duplicated_regions[\"CG_frac_r\"][duplicated_regions[\"CG_frac_l\"]==0.0]   #np.mean(duplicated_regions[\"CG_frac_l\"])\n",
    "    \n",
    "    duplicated_regions[\"CG_frac_r\"][duplicated_regions[\"CG_frac_r\"]==-1.000]=0.0\n",
    "    duplicated_regions[\"CG_frac_r\"][duplicated_regions[\"CG_frac_r\"]==0.0]=duplicated_regions[\"CG_frac_l\"][duplicated_regions[\"CG_frac_r\"]==0.0] #np.mean(duplicated_regions[\"CG_frac_r\"])\n",
    "    \n",
    "    duplicated_regions[\"CG_frac_in\"][duplicated_regions[\"CG_frac_in\"]==-1.000]=0.0\n",
    "    duplicated_regions[\"CG_frac_in\"][duplicated_regions[\"CG_frac_in\"]==0.0]=np.mean(duplicated_regions[\"CG_frac_in\"])\n",
    "    duplicated_regions[\"CG_frac\"]=(duplicated_regions[\"CG_frac_l\"]+duplicated_regions[\"CG_frac_r\"])/2\n",
    "    \n",
    "    duplicated_regions[\"CG_frac\"][duplicated_regions[\"CG_frac\"]==0.0]=np.mean(duplicated_regions[\"CG_frac_in\"])\n",
    "    \n",
    "    duplicated_regions.drop([\"CG_frac_l\",\"CG_frac_r\"],axis=1,inplace=True)\n",
    "    \n",
    "    for i in add_columns:\n",
    "        duplicated_regions[i]=old_data[i]\n",
    "    left_right_columns=[\"DNA\",\"LINE\",\"LTR\",\"SINE\",\"Low_complexity\",\"Retroposon\",\"Satellite\",\"Simple_repeat\",\"rRNA\",\"snRNA\",\"scRNA\",\"srpRNA\",\"tRNA\",\"RC\", 'L1_s', 'L2_s', 'MIR_s', 'Alu_s', 'Satellite_s']\n",
    "    \n",
    "    for i in left_right_columns:\n",
    "        duplicated_regions[i]=duplicated_regions[i+\"_r\"]+duplicated_regions[i+\"_l\"]\n",
    "        duplicated_regions.drop([i+\"_r\",i+\"_l\"],axis=1,inplace=True)\n",
    "        \n",
    "    return duplicated_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1520dbbe-1fb9-4306-beef-609ceb5e25c4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "ed920c8c-9022-4f46-a513-b365f7cea3ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_duplicated_regions(splits,duplicated_regions,chromosome_name):\n",
    "    \n",
    "    splits2=copy.deepcopy(splits)\n",
    "    df=duplicated_regions[duplicated_regions[\"chr\"]==chromosome_name]\n",
    "    for i,row in df.iterrows():\n",
    "        splits2[row[\"coor_s\"]:row[\"coor_e\"]]=row[\"ids\"]\n",
    "        \n",
    "    interval_with_dupliated_regions={}\n",
    "    for i in splits:\n",
    "        overlaps=splits2.overlap(i.begin,i.end)\n",
    "        overlaps.remove(Interval(i.begin,i.end,(i.begin,i.end)))\n",
    "        interval_with_dupliated_regions[(i.begin,i.end)]=overlaps\n",
    "        \n",
    "    return interval_with_dupliated_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "7e20c3bf-2dea-471a-b880-a1369927b0be",
   "metadata": {},
   "outputs": [],
   "source": [
    "def length_of_overlap(reg1,reg2):\n",
    "    if reg1[0]<=reg2[0]:\n",
    "        return(reg1[1]-reg2[0])\n",
    "    else:\n",
    "        return(reg2[1]-reg1[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "54ac0281-90f8-45ea-9137-2178522d47a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def binning(x):\n",
    "    if x>0.933:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "d1451b72-967c-45cc-8023-20012e25cd2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def binning1(x): #(1.745, 0.196)\n",
    "    if x>0.40174999999999994:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "61c391b3-f3cf-4001-a1f3-a866410d8e71",
   "metadata": {},
   "outputs": [],
   "source": [
    "def binning2(x): #0.749, 0.065\n",
    "    if x>0.424:\n",
    "        return 1\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "93bc463b-f7c6-49b8-bcbb-7129448be14a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def binning_length(x):\n",
    "    if x<2574.5:\n",
    "        return 0 \n",
    "    else:\n",
    "        return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "25df1460-e9f2-454c-8410-dce860466ef5",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns_needed=[\"chromosome_name\",\"start\", \"end\",'length_0','length_1', 'jumps',  'gaps', 'genes',\n",
    "       'cpgisl_in', 'cpgisl_bor', 'repli_in', 'repli_bor', 'repli_bor_deriv',\n",
    "       'repli_deriv', 'recomb_in', 'recomb_bor', 'dnase_in', 'dnase_bor','DNA', 'LINE',\n",
    "       'LTR', 'SINE', 'Low_complexity', 'Retroposon', 'Satellite',\n",
    "       'Simple_repeat', 'rRNA', 'snRNA', 'scRNA', 'srpRNA', 'tRNA', 'RC',\n",
    "       'L1_s', 'L2_s', 'MIR_s', 'Alu_s', 'Satellite_s','component_size', 'intra_degree', 'iner_degree', 'self_loops',\n",
    "       'edges_double', 'edges_tandem','edges_ident_mean_0','edges_ident_mean_1',\"CG_frac_0\",\"CG_frac_1\",\"CG_frac_in_0\",\"CG_frac_in_1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "b0eb6555-2032-4b56-88b7-36686045c6e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(columns_needed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "2b9d353f-11a0-434d-8cc3-7d7b8616f1cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CG_frac_0'"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns_needed[45]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "dc95c182-1782-4e78-8116-603b2b31dc1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_training_data(duplicated_regions):\n",
    "    \n",
    "    columns_needed=[\"chromosome_name\",\"start\", \"end\",'length_0','length_1', 'jumps',  'gaps', 'genes',\n",
    "       'cpgisl_in', 'cpgisl_bor', 'repli_in', 'repli_bor', 'repli_bor_deriv',\n",
    "       'repli_deriv', 'recomb_in', 'recomb_bor', 'dnase_in', 'dnase_bor','DNA', 'LINE',\n",
    "       'LTR', 'SINE', 'Low_complexity', 'Retroposon', 'Satellite',\n",
    "       'Simple_repeat', 'rRNA', 'snRNA', 'scRNA', 'srpRNA', 'tRNA', 'RC',\n",
    "       'L1_s', 'L2_s', 'MIR_s', 'Alu_s', 'Satellite_s','component_size', 'intra_degree', 'iner_degree', 'self_loops',\n",
    "       'edges_double', 'edges_tandem','edges_ident_mean_0','edges_ident_mean_1',\"CG_frac_0\",\"CG_frac_1\",\"CG_frac_in_0\",\"CG_frac_in_1\"]\n",
    "    \n",
    "    df=pd.DataFrame(columns=columns_needed)\n",
    "    \n",
    "    \n",
    "    for i in range(1,23):\n",
    "        print(i)\n",
    "        chr_name=\"chr\"+str(i)\n",
    "        splits=get_splits(chr_name)\n",
    "        overlap_data=find_duplicated_regions(splits,duplicated_regions,i)\n",
    "\n",
    "\n",
    "        for split in overlap_data:\n",
    "            start=split[0]\n",
    "            end=split[1]\n",
    "            cumm_features=np.zeros(len(columns_needed))\n",
    "            # START AND END POSITION OF SPLIT\n",
    "            cumm_features[0]=i\n",
    "            cumm_features[1]=start\n",
    "            cumm_features[2]=end\n",
    "\n",
    "            for dups in overlap_data[split]:\n",
    "\n",
    "                # EXTRACTING RELEVENT ROW FROM FEATURE DATA IE. DUPLICATION REGION.\n",
    "                dups_id=dups[2]\n",
    "                dup_data=features[features[\"ids\"]==dups_id]\n",
    "                \n",
    "\n",
    "                # FOR EXTRACTING OVERLAP LENGTH RATIO \n",
    "                if (len(dup_data))!=0:\n",
    "                    length_of_dup=np.array(dup_data[\"length\"])[0]\n",
    "                    cumm_features[3+binning_length(length_of_dup)]+=1\n",
    "                    # REPLI* , RECOMB* , DNASE* FEATURES \n",
    "                    index_repli=5\n",
    "                    for count in np.array(dup_data[['jumps',  'gaps', 'genes',\n",
    "               'cpgisl_in', 'cpgisl_bor', 'repli_in', 'repli_bor', 'repli_bor_deriv',\n",
    "               'repli_deriv', 'recomb_in', 'recomb_bor', 'dnase_in', 'dnase_bor','DNA', 'LINE',\n",
    "               'LTR', 'SINE', 'Low_complexity', 'Retroposon', 'Satellite',\n",
    "               'Simple_repeat', 'rRNA', 'snRNA', 'scRNA', 'srpRNA', 'tRNA', 'RC',\n",
    "               'L1_s', 'L2_s', 'MIR_s', 'Alu_s', 'Satellite_s','component_size', 'intra_degree', 'iner_degree', 'self_loops',\n",
    "               'edges_double', 'edges_tandem',\"CG_frac\",\"CG_frac_in\"]])[0]:\n",
    "                        cumm_features[index_repli]+=int(count)\n",
    "                        index_repli+=1\n",
    "                    #EDGES IDENT MEAN\n",
    "                    index_edges_ident_mean=43\n",
    "                    cumm_features[index_edges_ident_mean + int(binning(np.array(dup_data['edges_ident_mean'])[0]))] += 1 \n",
    "                    \n",
    "                    index_cg_frac=45\n",
    "                    \n",
    "                    cumm_features[index_cg_frac + int(binning1(np.array(dup_data['CG_frac'])[0]))] += 1\n",
    "                    \n",
    "                    index_cg_frac_in=47\n",
    "                    \n",
    "                    cumm_features[index_cg_frac_in + int(binning2(np.array(dup_data['CG_frac_in'])[0]))] += 1\n",
    "                    \n",
    "            # ADDING ROW DATA INTO DATAFRAME\n",
    "\n",
    "            df2 = pd.DataFrame([cumm_features],\n",
    "                       columns=columns_needed)\n",
    "\n",
    "            df=pd.concat([df, df2])\n",
    "    \n",
    "        \n",
    "        \n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "dcd87fdc-ae81-4a0e-97c1-bd967725bb15",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/local/ipykernel_45272/1614765331.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  duplicated_regions[\"CG_frac_l\"][duplicated_regions[\"CG_frac_l\"]==-1.000]=0.0\n",
      "/scratch/local/ipykernel_45272/1614765331.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  duplicated_regions[\"CG_frac_l\"][duplicated_regions[\"CG_frac_l\"]==0.0]=duplicated_regions[\"CG_frac_r\"][duplicated_regions[\"CG_frac_l\"]==0.0]   #np.mean(duplicated_regions[\"CG_frac_l\"])\n",
      "/scratch/local/ipykernel_45272/1614765331.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  duplicated_regions[\"CG_frac_r\"][duplicated_regions[\"CG_frac_r\"]==-1.000]=0.0\n",
      "/scratch/local/ipykernel_45272/1614765331.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  duplicated_regions[\"CG_frac_r\"][duplicated_regions[\"CG_frac_r\"]==0.0]=duplicated_regions[\"CG_frac_l\"][duplicated_regions[\"CG_frac_r\"]==0.0] #np.mean(duplicated_regions[\"CG_frac_r\"])\n",
      "/scratch/local/ipykernel_45272/1614765331.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  duplicated_regions[\"CG_frac_in\"][duplicated_regions[\"CG_frac_in\"]==-1.000]=0.0\n",
      "/scratch/local/ipykernel_45272/1614765331.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  duplicated_regions[\"CG_frac_in\"][duplicated_regions[\"CG_frac_in\"]==0.0]=np.mean(duplicated_regions[\"CG_frac_in\"])\n",
      "/scratch/local/ipykernel_45272/1614765331.py:18: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  duplicated_regions[\"CG_frac\"][duplicated_regions[\"CG_frac\"]==0.0]=np.mean(duplicated_regions[\"CG_frac_in\"])\n"
     ]
    }
   ],
   "source": [
    "features=get_duplicated_regions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "1fb8f8e1-90d1-4fd5-ac42-a78e2533d525",
   "metadata": {},
   "outputs": [],
   "source": [
    "features.to_csv(\"Duplicated_Regions_6.6k.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "5cb02ee8-bb91-46eb-b379-9e151761a0f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chr', 'coor_s', 'coor_e', 'ids', 'jumps', 'length', 'centro', 'telo',\n",
       "       'gaps', 'genes', 'intra_frac', 'cpgisl_in', 'cpgisl_bor', 'ctcf',\n",
       "       'repli_in', 'repli_bor', 'repli_bor_deriv', 'repli_deriv', 'repli_vari',\n",
       "       'recomb_in', 'recomb_bor', 'dnase_in', 'dnase_bor', 'used_coor_l_s',\n",
       "       'used_coor_l_e', 'used_coor_r_s', 'used_coor_r_e', 'CG_frac_in',\n",
       "       'CG_frac', 'component_size', 'intra_degree', 'iner_degree',\n",
       "       'self_loops', 'edges_double', 'edges_tandem', 'edges_ident_mean', 'DNA',\n",
       "       'LINE', 'LTR', 'SINE', 'Low_complexity', 'Retroposon', 'Satellite',\n",
       "       'Simple_repeat', 'rRNA', 'snRNA', 'scRNA', 'srpRNA', 'tRNA', 'RC',\n",
       "       'L1_s', 'L2_s', 'MIR_s', 'Alu_s', 'Satellite_s'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a4e4676-0e53-4360-a977-e9a6435f00b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "45364466-0a5f-4e2a-ad3a-16994e4cf0c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.424"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median(features['CG_frac_in'])  #features[\"CG_frac_in\"]),np.min(features[\"CG_frac_in\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "0a65b87c-9fbc-4782-95e0-397ea73febe0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1.0, 0.098)"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(features[\"CG_frac\"]),np.min(features[\"CG_frac\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "5d42f36e-ef71-41ed-bd44-900b3399eac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "3\n",
      "2\n",
      "3\n",
      "3\n",
      "3\n",
      "4\n",
      "3\n",
      "5\n",
      "3\n",
      "6\n",
      "3\n",
      "7\n",
      "3\n",
      "8\n",
      "3\n",
      "9\n",
      "3\n",
      "10\n",
      "3\n",
      "11\n",
      "3\n",
      "12\n",
      "3\n",
      "13\n",
      "3\n",
      "14\n",
      "3\n",
      "15\n",
      "3\n",
      "16\n",
      "3\n",
      "17\n",
      "3\n",
      "18\n",
      "3\n",
      "19\n",
      "3\n",
      "20\n",
      "3\n",
      "21\n",
      "3\n",
      "22\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "df_ML_output=create_training_data(features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "43a11d2d-c6b7-4cea-a903-b32694a4f03a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length_0</th>\n",
       "      <th>length_1</th>\n",
       "      <th>jumps</th>\n",
       "      <th>gaps</th>\n",
       "      <th>genes</th>\n",
       "      <th>cpgisl_in</th>\n",
       "      <th>cpgisl_bor</th>\n",
       "      <th>...</th>\n",
       "      <th>iner_degree</th>\n",
       "      <th>self_loops</th>\n",
       "      <th>edges_double</th>\n",
       "      <th>edges_tandem</th>\n",
       "      <th>edges_ident_mean_0</th>\n",
       "      <th>edges_ident_mean_1</th>\n",
       "      <th>CG_frac_0</th>\n",
       "      <th>CG_frac_1</th>\n",
       "      <th>CG_frac_in_0</th>\n",
       "      <th>CG_frac_in_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>10001.0</td>\n",
       "      <td>10010001.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>140.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>218946419.0</td>\n",
       "      <td>228946419.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>30.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>165100005.0</td>\n",
       "      <td>175100005.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>24.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>125100001.0</td>\n",
       "      <td>135100001.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>62.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>30010004.0</td>\n",
       "      <td>40010004.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22.0</td>\n",
       "      <td>10001.0</td>\n",
       "      <td>10010001.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22.0</td>\n",
       "      <td>17400001.0</td>\n",
       "      <td>27400001.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>166.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>107.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>...</td>\n",
       "      <td>77.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>255.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>84.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22.0</td>\n",
       "      <td>27400002.0</td>\n",
       "      <td>37400002.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22.0</td>\n",
       "      <td>40808467.0</td>\n",
       "      <td>50808467.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22.0</td>\n",
       "      <td>30808466.0</td>\n",
       "      <td>40808466.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>301 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome_name        start          end  length_0  length_1  jumps  \\\n",
       "0               1.0      10001.0   10010001.0       7.0      29.0   51.0   \n",
       "0               1.0  218946419.0  228946419.0      13.0      15.0   36.0   \n",
       "0               1.0  165100005.0  175100005.0      14.0       6.0   24.0   \n",
       "0               1.0  125100001.0  135100001.0       0.0       4.0    9.0   \n",
       "0               1.0   30010004.0   40010004.0      25.0       8.0   35.0   \n",
       "..              ...          ...          ...       ...       ...    ...   \n",
       "0              22.0      10001.0   10010001.0       0.0       0.0    0.0   \n",
       "0              22.0   17400001.0   27400001.0      23.0      86.0  166.0   \n",
       "0              22.0   27400002.0   37400002.0      13.0      19.0   39.0   \n",
       "0              22.0   40808467.0   50808467.0      24.0      17.0   51.0   \n",
       "0              22.0   30808466.0   40808466.0       9.0      19.0   36.0   \n",
       "\n",
       "    gaps  genes  cpgisl_in  cpgisl_bor  ...  iner_degree  self_loops  \\\n",
       "0    9.0   40.0       44.0         2.0  ...        140.0        11.0   \n",
       "0    2.0   15.0       44.0         2.0  ...         30.0         2.0   \n",
       "0    0.0   15.0        1.0         0.0  ...         24.0         0.0   \n",
       "0    4.0    0.0        2.0         1.0  ...         62.0         3.0   \n",
       "0    0.0   23.0        7.0         2.0  ...         39.0         0.0   \n",
       "..   ...    ...        ...         ...  ...          ...         ...   \n",
       "0    0.0    0.0        0.0         0.0  ...          0.0         0.0   \n",
       "0    6.0  107.0       69.0         4.0  ...         77.0        10.0   \n",
       "0    0.0   25.0        3.0         1.0  ...          7.0         2.0   \n",
       "0    1.0   34.0       12.0         1.0  ...         19.0         0.0   \n",
       "0    0.0   26.0        4.0         2.0  ...         15.0         3.0   \n",
       "\n",
       "    edges_double  edges_tandem  edges_ident_mean_0  edges_ident_mean_1  \\\n",
       "0          184.0          28.0                13.0                24.0   \n",
       "0           96.0          10.0                17.0                11.0   \n",
       "0            0.0           2.0                12.0                 8.0   \n",
       "0          101.0           2.0                 2.0                 2.0   \n",
       "0            0.0           9.0                16.0                17.0   \n",
       "..           ...           ...                 ...                 ...   \n",
       "0            0.0           0.0                 0.0                 0.0   \n",
       "0          255.0          80.0                38.0                71.0   \n",
       "0            2.0           4.0                14.0                18.0   \n",
       "0            8.0          24.0                19.0                22.0   \n",
       "0            4.0           8.0                12.0                16.0   \n",
       "\n",
       "    CG_frac_0  CG_frac_1  CG_frac_in_0  CG_frac_in_1  \n",
       "0         5.0       31.0           6.0          30.0  \n",
       "0        13.0       15.0          13.0          15.0  \n",
       "0        16.0        4.0          14.0           6.0  \n",
       "0         1.0        3.0           4.0           0.0  \n",
       "0        11.0       22.0           8.0          25.0  \n",
       "..        ...        ...           ...           ...  \n",
       "0         0.0        0.0           0.0           0.0  \n",
       "0        25.0       84.0          25.0          84.0  \n",
       "0         7.0       25.0          11.0          21.0  \n",
       "0        11.0       30.0           6.0          35.0  \n",
       "0         5.0       23.0           9.0          19.0  \n",
       "\n",
       "[301 rows x 49 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ML_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "d5071f4b-87e3-4719-ae06-189c8ef1d166",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome_name', 'start', 'end', 'length_0', 'length_1', 'jumps',\n",
       "       'gaps', 'genes', 'cpgisl_in', 'cpgisl_bor', 'repli_in', 'repli_bor',\n",
       "       'repli_bor_deriv', 'repli_deriv', 'recomb_in', 'recomb_bor', 'dnase_in',\n",
       "       'dnase_bor', 'DNA', 'LINE', 'LTR', 'SINE', 'Low_complexity',\n",
       "       'Retroposon', 'Satellite', 'Simple_repeat', 'rRNA', 'snRNA', 'scRNA',\n",
       "       'srpRNA', 'tRNA', 'RC', 'L1_s', 'L2_s', 'MIR_s', 'Alu_s', 'Satellite_s',\n",
       "       'component_size', 'intra_degree', 'iner_degree', 'self_loops',\n",
       "       'edges_double', 'edges_tandem', 'edges_ident_mean_0',\n",
       "       'edges_ident_mean_1', 'CG_frac_0', 'CG_frac_1', 'CG_frac_in_0',\n",
       "       'CG_frac_in_1'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ML_output.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "31d4ce8b-ad45-4bd4-8f16-889bc546e41a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     5.0\n",
       "0    13.0\n",
       "0    16.0\n",
       "0     1.0\n",
       "0    11.0\n",
       "     ... \n",
       "0     0.0\n",
       "0    25.0\n",
       "0     7.0\n",
       "0    11.0\n",
       "0     5.0\n",
       "Name: CG_frac_0, Length: 301, dtype: float64"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ML_output[\"CG_frac_0\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "f018cfb7-c448-4c32-87f4-bb8e22dbdd99",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ML_output.to_csv(\"Duplicated_Regions_Final_10MB.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "e355b3ab-2f05-415c-bf42-a1d2fcade84d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>coor_s</th>\n",
       "      <th>coor_e</th>\n",
       "      <th>ids</th>\n",
       "      <th>jumps</th>\n",
       "      <th>length</th>\n",
       "      <th>centro</th>\n",
       "      <th>telo</th>\n",
       "      <th>gaps</th>\n",
       "      <th>genes</th>\n",
       "      <th>...</th>\n",
       "      <th>MIR_s_r</th>\n",
       "      <th>Alu_s_r</th>\n",
       "      <th>Satellite_s_r</th>\n",
       "      <th>used_coor_l_s</th>\n",
       "      <th>used_coor_l_e</th>\n",
       "      <th>used_coor_r_s</th>\n",
       "      <th>used_coor_r_e</th>\n",
       "      <th>CG_frac_l</th>\n",
       "      <th>CG_frac_r</th>\n",
       "      <th>CG_frac_in</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>817367</td>\n",
       "      <td>821400</td>\n",
       "      <td>id5</td>\n",
       "      <td>1</td>\n",
       "      <td>4033</td>\n",
       "      <td>121205059</td>\n",
       "      <td>807367</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>817317</td>\n",
       "      <td>817367</td>\n",
       "      <td>821400</td>\n",
       "      <td>821450</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.529</td>\n",
       "      <td>0.510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6651</th>\n",
       "      <td>22</td>\n",
       "      <td>48911634</td>\n",
       "      <td>48912886</td>\n",
       "      <td>id4359</td>\n",
       "      <td>1</td>\n",
       "      <td>1252</td>\n",
       "      <td>33857316</td>\n",
       "      <td>1895582</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>48911584</td>\n",
       "      <td>48911634</td>\n",
       "      <td>48912886</td>\n",
       "      <td>48912936</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0.451</td>\n",
       "      <td>0.468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6652</th>\n",
       "      <td>22</td>\n",
       "      <td>49383944</td>\n",
       "      <td>49385910</td>\n",
       "      <td>id4360</td>\n",
       "      <td>1</td>\n",
       "      <td>1966</td>\n",
       "      <td>34329626</td>\n",
       "      <td>1422558</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49383894</td>\n",
       "      <td>49383944</td>\n",
       "      <td>49385910</td>\n",
       "      <td>49385960</td>\n",
       "      <td>0.510</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6653</th>\n",
       "      <td>22</td>\n",
       "      <td>49386637</td>\n",
       "      <td>49388496</td>\n",
       "      <td>id4361</td>\n",
       "      <td>1</td>\n",
       "      <td>1859</td>\n",
       "      <td>34332319</td>\n",
       "      <td>1419972</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49386587</td>\n",
       "      <td>49386637</td>\n",
       "      <td>49388496</td>\n",
       "      <td>49388546</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6654</th>\n",
       "      <td>22</td>\n",
       "      <td>50432257</td>\n",
       "      <td>50442552</td>\n",
       "      <td>id4362</td>\n",
       "      <td>2</td>\n",
       "      <td>10295</td>\n",
       "      <td>35377939</td>\n",
       "      <td>365916</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>50432207</td>\n",
       "      <td>50432257</td>\n",
       "      <td>50442552</td>\n",
       "      <td>50442602</td>\n",
       "      <td>0.725</td>\n",
       "      <td>0.451</td>\n",
       "      <td>0.648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6655</th>\n",
       "      <td>22</td>\n",
       "      <td>50740515</td>\n",
       "      <td>50808468</td>\n",
       "      <td>id4363</td>\n",
       "      <td>2</td>\n",
       "      <td>67953</td>\n",
       "      <td>35686197</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>50740465</td>\n",
       "      <td>50740515</td>\n",
       "      <td>50808468</td>\n",
       "      <td>50808518</td>\n",
       "      <td>0.569</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6656 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr    coor_s    coor_e     ids  jumps  length     centro     telo  \\\n",
       "0       0         0         0     0.0      0       0          0        0   \n",
       "1       0         0         0     0.0      0       0          0        0   \n",
       "2       0         0         0     0.0      0       0          0        0   \n",
       "3       0         0         0     0.0      0       0          0        0   \n",
       "4       1    817367    821400     id5      1    4033  121205059   807367   \n",
       "...   ...       ...       ...     ...    ...     ...        ...      ...   \n",
       "6651   22  48911634  48912886  id4359      1    1252   33857316  1895582   \n",
       "6652   22  49383944  49385910  id4360      1    1966   34329626  1422558   \n",
       "6653   22  49386637  49388496  id4361      1    1859   34332319  1419972   \n",
       "6654   22  50432257  50442552  id4362      2   10295   35377939   365916   \n",
       "6655   22  50740515  50808468  id4363      2   67953   35686197        0   \n",
       "\n",
       "      gaps  genes  ...  MIR_s_r  Alu_s_r  Satellite_s_r  used_coor_l_s  \\\n",
       "0        0      0  ...        0        0              0              0   \n",
       "1        0      0  ...        0        0              0              0   \n",
       "2        0      0  ...        0        0              0              0   \n",
       "3        0      0  ...        0        0              0              0   \n",
       "4        0      1  ...        0        0              0         817317   \n",
       "...    ...    ...  ...      ...      ...            ...            ...   \n",
       "6651     0      0  ...        0        0              0       48911584   \n",
       "6652     0      0  ...        0        0              0       49383894   \n",
       "6653     0      0  ...        0        0              0       49386587   \n",
       "6654     0      1  ...        0        1              0       50432207   \n",
       "6655     1      3  ...        0        0              0       50740465   \n",
       "\n",
       "      used_coor_l_e  used_coor_r_s  used_coor_r_e  CG_frac_l  CG_frac_r  \\\n",
       "0                 0              0              0      0.000      0.000   \n",
       "1                 0              0              0      0.000      0.000   \n",
       "2                 0              0              0      0.000      0.000   \n",
       "3                 0              0              0      0.000      0.000   \n",
       "4            817367         821400         821450      0.392      0.529   \n",
       "...             ...            ...            ...        ...        ...   \n",
       "6651       48911634       48912886       48912936      0.490      0.451   \n",
       "6652       49383944       49385910       49385960      0.510      0.392   \n",
       "6653       49386637       49388496       49388546      0.392      0.314   \n",
       "6654       50432257       50442552       50442602      0.725      0.451   \n",
       "6655       50740515       50808468       50808518      0.569      0.000   \n",
       "\n",
       "      CG_frac_in  \n",
       "0          0.000  \n",
       "1          0.000  \n",
       "2          0.000  \n",
       "3          0.000  \n",
       "4          0.510  \n",
       "...          ...  \n",
       "6651       0.468  \n",
       "6652       0.393  \n",
       "6653       0.397  \n",
       "6654       0.648  \n",
       "6655       0.454  \n",
       "\n",
       "[6656 rows x 68 columns]"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicated_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "3d501c4f-9440-4bc5-9a59-aa2499e3b052",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD8CAYAAACMwORRAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy89olMNAAAACXBIWXMAAAsTAAALEwEAmpwYAAAP4klEQVR4nO3db4xcV33G8e9DoKECRKDeWsZxuoi6tKESpl2lqWgll4iSBFQDomlCFQyyalolEqi8wPAGKjVSkAoRhTataSInFRAsCIrbRG3TFBRQy59NGoUkLq0LjmLLxIaEAEJAbX59sdfJxN71zuzM7Oye/X6kke+ce+bOb89unntz5t47qSokSW15xqQLkCSNnuEuSQ0y3CWpQYa7JDXIcJekBhnuktSgZ066AIB169bV9PT0pMuQpFXlnnvu+XZVTc23bkWE+/T0NLOzs5MuQ5JWlSQPL7TOaRlJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSg1bERUyjMr3r9qc9P3jtaydUiSRNlkfuktQgw12SGmS4S1KDDHdJapDhLkkNWvRsmSTPBu4Gzu76f7qq3pfkxcAtwM8B9wBXVtVPkpwN3Az8OvAd4A+q6uCY6u9b75k0nkUjqXX9HLn/GHhVVb0c2AJcnORC4APAdVX1i8DjwI6u/w7g8a79uq6fJGkZLRruNecH3dNndY8CXgV8umu/CXh9t7yte063/qIkGVXBkqTF9TXnnuSsJPcBR4E7gf8FvltVx7suh4CN3fJG4BGAbv0TzE3dSJKWSV9XqFbVCWBLknOAzwK/POwbJ9kJ7AQ477zzht3cvE69YlWS1oqBzpapqu8CnwN+Ezgnycmdw7nA4W75MLAJoFv/fOY+WD11W7uraqaqZqam5v1+V0nSEi0a7kmmuiN2kvws8GpgP3Mh/6au23bgtm55X/ecbv2/VVWNsOZlMb3r9icfkrTa9DMtswG4KclZzO0M9lbVPyZ5CLglyZ8D/wnc0PW/Afj7JAeAx4DLx1C3JOkMFg33qrofeMU87d8ALpin/UfA74+kujHxnHdJrfMKVUlq0Kq/n7tz4pJ0Oo/cJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lq0Ko/z31YXq0qqUUeuUtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGrflTIXt5+2BJrfDIXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBi0a7kk2JflckoeSPJjkHV37+5McTnJf97i05zXvSXIgydeTvGacP4Ak6XT9XMR0HHhXVd2b5HnAPUnu7NZdV1V/0ds5yfnA5cDLgBcB/5rkl6rqxCgLlyQtbNEj96o6UlX3dsvfB/YDG8/wkm3ALVX146r6JnAAuGAUxUqS+jPQnHuSaeAVwJe7pquT3J/kxiQv6No2Ao/0vOwQZ94ZSJJGrO9wT/Jc4DPAO6vqe8D1wEuALcAR4IODvHGSnUlmk8weO3ZskJdKkhbRV7gneRZzwf7xqroVoKoeraoTVfVT4GM8NfVyGNjU8/Jzu7anqardVTVTVTNTU1PD/AySpFP0c7ZMgBuA/VX1oZ72DT3d3gA80C3vAy5PcnaSFwObga+MrmRJ0mL6OVvmlcCVwNeS3Ne1vRe4IskWoICDwNsBqurBJHuBh5g70+Yqz5SRpOW1aLhX1ReBzLPqjjO85hrgmiHqkiQNwStUJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoP6Oc99zZvedfuTywevfe0EK5Gk/njkLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhpkuEtSgzzPfUCe8y5pNfDIXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGrRouCfZlORzSR5K8mCSd3TtL0xyZ5L/6f59QdeeJH+Z5ECS+5P82rh/CEnS0/Vz5H4ceFdVnQ9cCFyV5HxgF3BXVW0G7uqeA1wCbO4eO4HrR161JOmMFg33qjpSVfd2y98H9gMbgW3ATV23m4DXd8vbgJtrzpeAc5JsGHXhkqSFDTTnnmQaeAXwZWB9VR3pVn0LWN8tbwQe6XnZoa5NkrRM+g73JM8FPgO8s6q+17uuqgqoQd44yc4ks0lmjx07NshLJUmL6CvckzyLuWD/eFXd2jU/enK6pfv3aNd+GNjU8/Jzu7anqardVTVTVTNTU1NLrV+SNI9+zpYJcAOwv6o+1LNqH7C9W94O3NbT/pburJkLgSd6pm8kScugny/reCVwJfC1JPd1be8FrgX2JtkBPAxc1q27A7gUOAD8EHjbKAuWJC1u0XCvqi8CWWD1RfP0L+CqIeuSJA3BK1QlqUF+h+oY+D2rkibNI3dJapDhLkkNMtwlqUHOuY9I7zy7JE2aR+6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgr1AdglelSlqpDPcx8/a/kibBaRlJapDhLkkNMtwlqUGGuyQ1yHCXpAYZ7pLUIMNdkhq0aLgnuTHJ0SQP9LS9P8nhJPd1j0t71r0nyYEkX0/ymnEVLklaWD9H7nuAi+dpv66qtnSPOwCSnA9cDryse81fJzlrVMVKkvqzaLhX1d3AY31ubxtwS1X9uKq+CRwALhiiPknSEgwz5351kvu7aZsXdG0bgUd6+hzq2k6TZGeS2SSzx44dG6IMSdKplhru1wMvAbYAR4APDrqBqtpdVTNVNTM1NbXEMiRJ81lSuFfVo1V1oqp+CnyMp6ZeDgOberqe27VJkpbRku4KmWRDVR3pnr4BOHkmzT7gE0k+BLwI2Ax8ZegqG+EdIiUtl0XDPcknga3AuiSHgPcBW5NsAQo4CLwdoKoeTLIXeAg4DlxVVSfGUrkkaUGLhntVXTFP8w1n6H8NcM0wRUmShuMVqpLUIMNdkhpkuEtSgwx3SWqQ4S5JDTLcJalBhrskNchwl6QGGe6S1CDDXZIaZLhLUoMMd0lqkOEuSQ0y3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDFv0OVS2v6V23P7l88NrXTrASSauZ4b4C9Aa6JI2C4T4hBrqkcTLcVzCnaCQtlR+oSlKDFj1yT3Ij8DrgaFX9atf2QuBTwDRwELisqh5PEuDDwKXAD4G3VtW94yl9bfEoXtIg+jly3wNcfErbLuCuqtoM3NU9B7gE2Nw9dgLXj6ZMSdIgFg33qrobeOyU5m3ATd3yTcDre9pvrjlfAs5JsmFEtUqS+rTUOff1VXWkW/4WsL5b3gg80tPvUNd2miQ7k8wmmT127NgSy5AkzWfoD1SrqoBawut2V9VMVc1MTU0NW4YkqcdSw/3Rk9Mt3b9Hu/bDwKaefud2bZKkZbTUcN8HbO+WtwO39bS/JXMuBJ7omb6RJC2Tfk6F/CSwFViX5BDwPuBaYG+SHcDDwGVd9zuYOw3yAHOnQr5tDDVLkhaxaLhX1RULrLponr4FXDVsUZKk4XiFqiQ1yHCXpAYZ7pLUIMNdkhpkuEtSg7yf+yq30Jd+eOdIaW0z3Fchv8VJ0mKclpGkBhnuktQgw12SGmS4S1KDDHdJapBny6xxfvG21CbDfQ3yVEqpfU7LSFKDDHdJapDhLkkNcs59DXCOXVp7DHcNZKGzazzrRlpZnJaRpAYZ7pLUIKdlGrXS59nHMY3j1JD0FMNdy2Yl7HCWsgNwp6HVyHDXk/r5sFTS6jBUuCc5CHwfOAEcr6qZJC8EPgVMAweBy6rq8eHKlCQNYhQfqP5OVW2pqpnu+S7grqraDNzVPZckLaNxTMtsA7Z2yzcBnwfePYb30YQtNF0zzBx1P1/47Ry4tLhhw72Af0lSwN9W1W5gfVUd6dZ/C1g/3wuT7AR2Apx33nlDliFNljscrTTDhvtvVdXhJD8P3Jnkv3pXVlV1wX+abkewG2BmZmbePlr9RvVh7DDb8QNhrUVDhXtVHe7+PZrks8AFwKNJNlTVkSQbgKMjqFPLbDkDcRzvZaBrrVvyB6pJnpPkeSeXgd8FHgD2Adu7btuB24YtUpI0mGGO3NcDn01ycjufqKp/SvJVYG+SHcDDwGXDlylJGsSSw72qvgG8fJ727wAXDVOUJGk4XqGqVc25dWl+hrvWrDPtGIY5ndHTIrUSeMtfSWqQR+7SPJzu0WpnuEsDcMpFq4XTMpLUIMNdkhrktIy0RP3MyzuNo0kx3KUJMPQ1boa71KB+7ouvthnu0irgkb4GZbhLy2Qc97b3i8y1EMNdmjCPyjUOhru0ggx69D2qb6hyp9Iez3OXpAZ55C7pjEfxnnmzOhnu0hrih65rh+Eu6Wn63QE4Z7+yOecuSQ3yyF3SRPl/AONhuEsa2qgCup/tuDPoj+EuaaQ8u2ZlMNwlLYtRXXC1ULs7j6cz3CWtSF59O5yxnS2T5OIkX09yIMmucb2PJOl0Ywn3JGcBfwVcApwPXJHk/HG8lyTpdOOalrkAOFBV3wBIcguwDXhoTO8naY3rZ15+XPqZ+lnuqaJxhftG4JGe54eA3+jtkGQnsLN7+qMkD86znecDTyzStg749lDVDm6+usa9jX77n6nfoOsc/8H7O/7j2cZKHX+AdfnAYL+DfGCQ3qfpreEXFuxVVSN/AG8C/q7n+ZXAR8/Qf3e/7ae2AbPj+BkW+fnmrXec2+i3/5n6DbrO8Xf8Hf++2pb1d9DvWIzrA9XDwKae5+d2bQv5hwHaF+q7nEZRw6Db6Lf/mfoNus7xH7y/4z+ebTj+T+mrhnR7gpFK8kzgv4GLmAv1rwJvrqr5pl6Gfa/ZqpoZ9XbVH8d/shz/yVupv4OxzLlX1fEkVwP/DJwF3DiOYO/sHtN21R/Hf7Ic/8lbkb+DsRy5S5Imy1v+SlKDDHdJapDhLkkNai7ck/xKkr9J8ukkfzLpetaiJM9JMpvkdZOuZa1JsjXJF7r/BrZOup61JskzklyT5CNJtk+yllUR7kluTHI0yQOntJ92c7Kq2l9VfwxcBrxyEvW2ZpDx77wb2Lu8VbZrwPEv4AfAs5m7MlxDGnD8tzF3Xc//MeHxXxXhDuwBLu5tONPNyZL8HnA7cMfyltmsPfQ5/klezdw9hI4ud5EN20P/f/9fqKpLmNvB/tky19mqPfQ//i8F/r2q/hSY6MzBqgj3qrobeOyU5idvTlZVPwFO3pyMqtrX/YH/4fJW2qYBx38rcCHwZuCPkqyKv7GVbJDxr6qfdusfB85exjKbNeDf/yHmxh7gxPJVebrV/GUd896crJtnfCNzf9geuY/PvONfVVcDJHkr8O2esNFoLfT3/0bgNcA5wEcnUNdasdDNET8MfCTJbwN3T6Kwk1ZzuM+rqj4PfH7CZax5VbVn0jWsRVV1K3DrpOtYq6rqh8COSdcBq2RaZgGD3pxMo+X4T5bjP1krfvxXc7h/Fdic5MVJfga4HNg34ZrWEsd/shz/yVrx478qwj3JJ4H/AF6a5FCSHVV1HDh5c7L9wN4x3pxsTXP8J8vxn6zVOv7eOEySGrQqjtwlSYMx3CWpQYa7JDXIcJekBhnuktQgw12SGmS4S1KDDHdJapDhLkkN+n/WKNA6unWHBgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plt.hist(features.length, bins=300)\n",
    "\n",
    "def plot_loghist(x, bins):\n",
    "  hist, bins = np.histogram(x, bins=bins)\n",
    "  logbins = np.logspace(np.log10(bins[0]),np.log10(bins[-1]),len(bins))\n",
    "  plt.hist(x, bins=logbins)\n",
    "  plt.xscale('log')\n",
    "\n",
    "plot_loghist(features.length, 100)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
